### PR TITLE
feat: make it support rewrite request body in external plugins

### PR DESCRIPTION
**Pull request by rubikplanet**
### Description

make it support rewrite request body.

please update ext-plugin-proto first.

to make it be able to rewrite the request body in external plugins.

i don't have a full development environment for apisix. someone have the full development env can help to add the test code.

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by jiangfucheng** [comment]
This PR is based on https://github.com/api7/ext-plugin-proto/pull/35, maybe we should wait for the PR to merge and then update the dependency version of `ext-plugin-proto`. And, we also need to add corresponding test.

File: apisix/plugins/ext-plugin/init.lua [hunk 1]
```diff
@@ -655,6 +655,12 @@ local rpc_handlers = {
                 end
             end
 
+            local len = rewrite:BodyLength()
+            if len > 0 then
+                let body = rewrite:BodyAsString()
+                ngx.req.set_body_data(body)
```
Comment (jiangfucheng) [inline apisix/plugins/ext-plugin/init.lua:661]:
```suggestion
                ngx.req.read_body()
                ngx.req.set_body_data(body)
```
We need call `ngx.req.read_body()` before `set_body_data`

**Comment by jiangfucheng** [comment]
https://github.com/apache/apisix/blob/59c1d99a6a571a77d9c3f7995a23785d93a4a92c/t/lib/ext-plugin.lua#L473C13-L473C13

Before this line, we can mock `rewrite_request_body` case:
```
        elseif case.rewrite_request_body == true then
            local len = 4
            http_req_call_rewrite.StartBodyVector(builder, len)
            builder:PrependByte(string.byte("\n"))
            builder:PrependByte(string.byte("c"))
            builder:PrependByte(string.byte("b"))
            builder:PrependByte(string.byte("a"))
            local b = builder:EndVector(len)

            http_req_call_rewrite.Start(builder)
            http_req_call_rewrite.AddBody(builder, b)
            local action = http_req_call_rewrite.End(builder)
            build_action(action, http_req_call_action.Rewrite)
```


And then, we can add test cases in `plugin/ext-plugin/http-req-call.t`
```
=== TEST 27: add route
--- config
    location /t {
        content_by_lua_block {
            local json = require("toolkit.json")
            local t = require("lib.test_admin")

            local code, message, res = t.test('/apisix/admin/routes/1',
                ngx.HTTP_PUT,
                 [[{
                    "uri": "/echo",
                    "plugins": {
                        "ext-plugin-pre-req": {
                        }
                    },
                    "upstream": {
                        "nodes": {
                            "127.0.0.1:1980": 1
                        },
                        "type": "roundrobin"
                    }
                }]]
            )

            if code >= 300 then
                ngx.status = code
                ngx.say(message)
                return
            end

            ngx.say(message)
        }
    }
--- response_body
passed



=== TEST 29: test rewrite request body
--- ONLY
--- request
GET /echo
--- response_body chomp
cat
--- extra_stream_config
    server {
        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock;

        content_by_lua_block {
            local ext = require("lib.ext-plugin")
            ext.go({rewrite_request_body = true})
        }
    }
--- response_body
abc
```

**Comment by jiangfucheng** [comment]
We aslo should update the documents.

**Comment by zxyao145** [comment]
Exciting feature, when can this PR be merged?  Many people need this feature

1. https://github.com/apache/apisix-java-plugin-runner/issues/143
1. https://github.com/apache/apisix-java-plugin-runner/issues/172
1. https://github.com/apache/apisix-java-plugin-runner/issues/222
1. https://github.com/apache/apisix-java-plugin-runner/issues/252
1. https://github.com/apache/apisix-go-plugin-runner/issues/127
1. https://github.com/apache/apisix-go-plugin-runner/issues/129

**Pull request closed without merging**